### DISCUSSION TITLE
How to use union metafunction to replace variant member and member functions that call std::visit.

### DISCUSSION THREAD

**Discussion by gregmarr** [hsutter/cppfront#766] [Suggestions]
This came up in a C++ study group meeting today about ways to replace virtual functions in a class hierarchy that has a fixed set of member classes.

Here is a solution from today using `std::variant` and `std::visit`.

```
#include <variant>

class Circle {
  double _radius = 0;

public:
  double GetArea() const { return 3.1415 * _radius * _radius; }
};

class Rectangle {
  double _width = 0;
  double _height = 0;

public:
  double GetArea() const { return _width * _height; }
};

class Shape {
  std::variant<Circle, Rectangle> _data;

public:
  double GetArea() const {
    return std::visit([](auto x) { return x.GetArea(); }, _data);
  }
};
```
One of the discussion points was that it would be nice to be able to get rid of the `std::visit` boilerplate and be able to automatically generate a "call this function on the active member of the variant" member function.

Thinking about how we could do this in Cpp2 with the union metafunction, I was able to get this far:

```
Circle: type = {
  _radius : double = 0;

  GetArea: () 3.1415 * _radius * _radius;
}

Rectangle: type = {
  _width : double = 0;
  _height : double = 0;

  GetArea: () _width * _height;
}

Shape: @union type = {
  circle: Circle;
  rectangle: Rectangle;

  // We can add member functions, but we have to write the type handling manually.
  // This is how we have to do it now:
  GetArea: (this) -> double = {
    if is_circle() { return circle().GetArea(); }
    else if is_rectangle() { return rectangle().GetArea(); }
    else { throw std::bad_variant_access(); }
  }
}
```

This is even worse in terms of having to do the boilerplate function, as you have to individually name the function.  You can't use `std::visit` because you don't have a `variant`, unless we make the union metafunction have all the hooks that visit uses from variant, which isn't out of the realm of possibility.  That would still require us to write the `std::visit` boilerplate.

Looking at the syntax in P0707 Section 4.2.2 `property metaclass function`, I think we could do something like this:

```
  visitor GetArea : (this) = {};
```

and the metafunction handler would rewrite the body of the visitor function like the above, similar to how it generates the destructor now.

**Comment by AbhinavK00** [comment]
Maybe the more experienced people here can explain better but I think pattern matching is the right answer to this use case.

**Comment by JohelEGP** [comment]
This looks like something you would build on top of `@union`.
`@union` isn't necessarily about similar data structures.
There must be a better name for this behavior.
The theme is common, though, so it's a reasonable extension
([one I'd benefit from](https://github.com/hsutter/cppfront/discussions/789#discussioncomment-7561840)).

**Reply by JohelEGP** [reply]
This reminds me of Rust traits, Carbon generics, and <https://github.com/ldionne/dyno>.

**Reply by AbhinavK00** [reply]
I was under the impression that `interfaces` would be usefu in these use cases but I haven't seen you guys use them. Is that because of they have virtual functions which just translates to runtime?

**Reply by JohelEGP** [reply]
> Is that because of they have virtual functions which just translates to runtime?

I think that's the answer.
It certainly would be for me.

**Reply by AbhinavK00** [reply]
Maybe that means the need for a metafunction that uses CRTP instead of virtual functions? 

Edit: But how would one generate the code on "derived" class?

**Reply by AbhinavK00** [reply]
I'm so sorry, I wrote the wrong acronym lol. I meant curiously recurring template pattern.

I probably looked stupid 😅

**Reply by JohelEGP** [reply]
With metafunctions, you don't even need templates in general.

There are two kind of data structures here.
Closed sets, like `std::variant`.
And open sets, like `@interface` and `@polymorphic_base` hierarchies.

All those can be implemented with type erasure.
For closed sets, you can also use static polymorphism.

**Reply by AbhinavK00** [reply]
Do you think something like the following is worth considering in cpp2?

It's a proposal in committee mailing that I think died in the mailing itself.

**Reply by AbhinavK00** [reply]
Sorry, it's a html file and I can't figure out how to share it here.

**Reply by JohelEGP** [reply]
Is it at <https://lists.isocpp.org/std-proposals/>?

**Reply by AbhinavK00** [reply]
Its this https://lists.isocpp.org/std-proposals/2023/06/6894.php

I don't think the proposal is in the discussion itself but I was more curious if this can be implemented in cpp2 at all.

**Reply by JohelEGP** [reply]
I see, it seems to be something similar.
Even if Cpp2 metafunctions isn't capable,
there's Cpp1 reflection: <https://www.reddit.com/r/cpp/comments/17q9dg4/abusing_the_c_preprocessor_for_declaring_rustlike/>.

**Reply by gregmarr** [reply]
> For closed sets, you can also use static polymorphism.

Yes, that's what this is about, implementing it within the union metatype instead of using a variant and writing boilerplate visitors.

**Reply by AbhinavK00** [reply]
> there's Cpp1 reflection: https://www.reddit.com/r/cpp/comments/17q9dg4/abusing_the_c_preprocessor_for_declaring_rustlike/.

Wow, that's very cool. Can't wait for cpp2 to have something like this. Btw, since there are so many "experiments" dealing with polymorphism using rust-like "thing" for cpp2, do you think it's worth it to have something like traits in cpp2?
I remember there was something from Sy Brand too related to this.

**Reply by JohelEGP** [reply]
If you can write a library for it in Cpp1,
you probably should be able to in Cpp2.

**Comment by AbhinavK00** [comment]
This is regarding the original problem posted by @gregmarr, maybe the `union` metafunction can have an active_member member function so that the GetArea function in your example could be written as:

```
GetArea: (this) -> double = {
    return this.active_member().GetArea();
  }
```

Specifics about this can be discusssed.

**Reply by JohelEGP** [reply]
This seems recursive.
`active_member()` returns a structure with a `GetArea()`,
but that's what you're trying to implement here.

**Reply by AbhinavK00** [reply]
But the member of the union have their own GetArea() which has been implemented. So, I don't think it's recursive. The example in [this reply](https://github.com/hsutter/cppfront/discussions/766#discussion-5758645)

**Reply by JohelEGP** [reply]
The thing is that `active_member().GetArea()` implies a data structure
with a function with the same implementation we're trying to implement.
That's because `active_member()` would need to return a type erased reference.
But that's what the `this` (a `@union`) already is.

**Reply by AbhinavK00** [reply]
You're right, so this idea was just wrong.
